lib/src/multipart_file.dart

```diff
@@ -4,108 +4,105 @@
-    contentType = contentType.change(parameters: {'charset': encoding.name});
-
-    return new MultipartFile.fromBytes(field, encoding.encode(value),
-        filename: filename,
-        contentType: contentType);
+    if (encoding != null) {
+      contentType = contentType.change(parameters: {'charset': encoding.name});
+    }
+
+    return new MultipartFile.fromStream(
+        field, new Stream<List<int>>.fromIterable([bytes]), bytes.length,
+        filename: filename, contentType: contentType);
   }
 
-  // TODO(nweiz): Infer the content-type from the filename.
-  /// Creates a new [MultipartFile] from a path to a file on disk.
-  ///
-  /// [filename] defaults to the basename of [filePath]. [contentType] currently
-  /// defaults to `application/octet-stream`, but in the future may be inferred
-  /// from [filename].
-  ///
-  /// Throws an [UnsupportedError] if `dart:io` isn't supported in this
-  /// environment.
-  static Future<MultipartFile> fromPath(String field, String filePath,
+  MultipartFile.fromStream(this.field, Stream<List<int>> stream, this.length,
+      {String filename, MediaType contentType})
+      : _stream = stream,
+        filename = filename,
+        contentType = contentType ??
+            _lookupMediaType(filename) ??
+            new MediaType('application', 'octet-stream');
+
+  static Future<MultipartFile> loadStream(
+      String field, Stream<List<int>> stream,
       {String filename, MediaType contentType}) async {
-    if (filename == null) filename = path.basename(filePath);
-    var file = new File(filePath);
-    var length = await file.length();
-    var stream = new ByteStream(DelegatingStream.typed(file.openRead()));
-    return new MultipartFile(field, stream, length,
-        filename: filename,
-        contentType: contentType);
+    var bytes = await collectBytes(stream);
+
+    return new MultipartFile(field, bytes,
+        filename: filename, contentType: contentType);
   }
 
-  // Finalizes the file in preparation for it being sent as part of a
-  // [MultipartRequest]. This returns a [ByteStream] that should emit the body
-  // of the file. The stream may be closed to indicate an empty file.
-  ByteStream finalize() {
-    if (isFinalized) {
-      throw new StateError("Can't finalize a finalized MultipartFile.");
+  /// Returns a [Stream] representing the contents of the file.
+  ///
+  /// Can only be called once.
+  Stream<List<int>> read() {
+    if (_stream == null) {
+      throw new StateError('The "read" method can only be called once on a '
+          'http.MultipartFile object.');
     }
-    _isFinalized = true;
-    return _stream;
+    var stream = _stream;
+    _stream = null;
+    return stream;
+  }
+
+  /// Looks up the [MediaType] from the [filename]'s extension or from
+  /// magic numbers contained within a file header's [bytes].
+  static MediaType _lookupMediaType(String filename, [List<int> bytes]) {
```

Comment:
Nit: `_lookUpMediaType`

"Lookup" is a word, but it's only a noun; the verb is always two words.